# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/HttpChannelParent.cpp
# Commit: 25822279ba5b
# Full Hash: 25822279ba5b7b695bd9173969877e8d5ba19000
# Author: Junior Hsu <juhsu@mozilla.com>
# Date: 2020-06-24 09:31:07
# Regressor Bug: 1633935
# File Overlap Count: 1
# Description:
#   Bug 1633935 - P12 top-level IPC stream filter should happen before OnStartRequest, r=mayhemer,necko-reviewers
#   
#   The WebRequest will suspend the channel in nsHttpChannel::PrepareToConnect(), apply the stream filter, send the main thread IPC  and resume.
#   That is, PHttpChannel::AttachStreamFilter should be handled in Child before OnStartRequest goes to listener.
#   
# ==============================================================================

diff -r e87b11c172b9 -r 25822279ba5b netwerk/protocol/http/HttpChannelParent.cpp
--- a/netwerk/protocol/http/HttpChannelParent.cpp	Tue Jun 23 16:58:38 2020 +0000
+++ b/netwerk/protocol/http/HttpChannelParent.cpp	Tue Jun 23 16:58:40 2020 +0000
@@ -83,7 +83,8 @@
       mCacheNeedFlowControlInitialized(false),
       mNeedFlowControl(true),
       mSuspendedForFlowControl(false),
-      mAfterOnStartRequestBegun(false) {
+      mAfterOnStartRequestBegun(false),
+      mStreamFilterAttached(false) {
   LOG(("Creating HttpChannelParent [this=%p]\n", this));
 
   // Ensure gHttpHandler is initialized: we need the atom table up and running.
@@ -1541,7 +1542,8 @@
   // Bug 1645901: Currently Set-Cookie is passed to child process on main
   // thread, which is racy with PBackground. We should have a way to set cookie
   // in child for Set-Cookie response header.
-  args.shouldWaitForOnStartRequestSent() = isDocument || hasSetCookie;
+  args.shouldWaitForOnStartRequestSent() =
+      isDocument || hasSetCookie || mStreamFilterAttached;
 
   rv = NS_OK;
 
@@ -2652,5 +2654,12 @@
   mOverrideReferrerInfo = aReferrerInfo;
 }
 
+bool HttpChannelParent::AttachStreamFilter(
+    Endpoint<extensions::PStreamFilterParent>&& aEndpoint) {
+  MOZ_ASSERT(!mAfterOnStartRequestBegun);
+  mStreamFilterAttached = true;
+  return SendAttachStreamFilter(std::move(aEndpoint));
+}
+
 }  // namespace net
 }  // namespace mozilla